### ISSUE TITLE
FP string_lit_as_bytes: match / slice patterns

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9885]
### Summary

.

### Lint Name

string_lit_as_bytes

### Reproducer

I tried this code:

```rust
#![feature(slice_patterns)]

fn main() {
    match "foo".as_bytes() {
        b"food" => println!("nom nom"),
        [b'f', ..] => println!("f!"),
        _ => ()
    }
}
```

I saw this happen:
`cargo clippy --fix -- -Wclippy::string_lit_as_bytes`
```
    Checking a v0.1.0 (/tmp/a)
warning: failed to automatically apply fixes suggested by rustc to crate `a`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
2 |     match b"foo" {
  |           ------ this expression has type `&[u8; 3]`
3 |         b"food" => println!("nom nom"),
  |         ^^^^^^^ expected an array with a fixed size of 3 elements, found one with 4 elements

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: calling `as_bytes()` on a string literal
 --> src/main.rs:2:11
  |
2 |     match "foo".as_bytes() {
  |           ^^^^^^^^^^^^^^^^ help: consider using a byte string literal instead: `b"foo"`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#string_lit_as_bytes
  = note: requested on the command line with `-W clippy::string-lit-as-bytes`

warning: `a` (bin "a") generated 1 warning (run `cargo fix --bin "a"` to apply 1 suggestion)
warning: `a` (bin "a" test) generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
```

### Version

```text
rustc 1.67.0-nightly (b833ad56f 2022-11-18)
binary: rustc
commit-hash: b833ad56f46a0bbe0e8729512812a161e7dae28a
commit-date: 2022-11-18
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
